File: src/add_city.py
"""
Add / Update City — Event Sources Discovery
--------------------------------------------
Discovers event sources for a city using available providers, merges the
results with any existing sources, then writes sources/{city_key}.yml.

Running for an existing city adds newly discovered sources without removing
any that are already there.

Usage (locally):
  ANTHROPIC_API_KEY=... CITY_NAME="Sydney, NSW, Australia" CITY_KEY=sydney python src/add_city.py
  PERPLEXITY_API_KEY=... CITY_NAME="Sydney, NSW, Australia" CITY_KEY=sydney python src/add_city.py

Usage (CI):
  Triggered by .github/workflows/add-city.yml
"""

import os
import re

import yaml

import anthropic_ai
import gemini_ai
import perplexity_ai

CITY_NAME = os.environ["CITY_NAME"]
CITY_KEY  = os.environ["CITY_KEY"]

SOURCES_DIR = "sources"
DIGEST_WF   = ".github/workflows/digest.yml"


def _extract_domain(source_str: str) -> str:
    """Return a normalised domain from a source string like 'Name (domain.com/path)'."""
    match = re.search(r'\(([^)]+)\)', source_str)
    if match:
        url = match.group(1).lstrip("https://").lstrip("http://").lstrip("www.")
        return url.split("/")[0].lower()
    return source_str.lower()


def merge_sources(target: dict, incoming: dict) -> None:
    """Add sources from incoming into target, deduplicating by domain."""
    for tier in ("aggregators", "institutions", "independents"):
        existing = {_extract_domain(s) for s in target.get(tier, [])}
        for source in incoming.get(tier, []):
            domain = _extract_domain(source)
            if domain not in existing:
                target[tier].append(source)
                existing.add(domain)


def load_existing_sources() -> dict:
    out_path = os.path.join(SOURCES_DIR, f"{CITY_KEY}.yml")
    if not os.path.exists(out_path):
        return {"aggregators": [], "institutions": [], "independents": []}
    with open(out_path) as f:
        data = yaml.safe_load(f) or {}
    src = data.get("sources", {})
    return {
        "aggregators":  list(src.get("aggregators",  [])),
        "institutions": list(src.get("institutions", [])),
        "independents": list(src.get("independents", [])),
    }


def discover_sources(existing: dict) -> dict:
    sources = {
        "aggregators":  list(existing["aggregators"]),
        "institutions": list(existing["institutions"]),
        "independents": list(existing["independents"]),
    }

    has_anthropic  = bool(os.environ.get("ANTHROPIC_API_KEY"))
    has_perplexity = bool(os.environ.get("PERPLEXITY_API_KEY"))
    has_gemini     = bool(os.environ.get("GOOGLE_API_KEY"))

    if not has_anthropic and not has_perplexity and not has_gemini:
        raise SystemExit(
            "✗ No API keys set (need at least one of ANTHROPIC_API_KEY, PERPLEXITY_API_KEY, GOOGLE_API_KEY)."
        )

    if has_anthropic:
        merge_sources(sources, anthropic_ai.find_sources(CITY_NAME))
    if has_perplexity:
        merge_sources(sources, perplexity_ai.find_sources(CITY_NAME))
    if has_gemini:
        merge_sources(sources, gemini_ai.find_sources(CITY_NAME))

    for tier in ("aggregators", "institutions", "independents"):
        before = len(existing[tier])
        after  = len(sources[tier])
        added  = after - before
        print(f"→ {tier}: {after} total ({f'+{added} new' if added else 'no new'})")
    return sources


def write_city_file(sources: dict) -> None:
    os.makedirs(SOURCES_DIR, exist_ok=True)
    out_path = os.path.join(SOURCES_DIR, f"{CITY_KEY}.yml")

    city_data = {
        "name":    CITY_NAME,
        "sources": {
            "aggregators":  sources.get("aggregators",  []),
            "institutions": sources.get("institutions", []),
            "independents": sources.get("independents", []),
        },
    }

    with open(out_path, "w") as f:
        yaml.dump(city_data, f, default_flow_style=False, allow_unicode=True, sort_keys=False)

    print(f"→ Written {out_path}")


def update_digest_workflow() -> None:
    with open(DIGEST_WF) as f:
        content = f.read()

    pattern = r'(        options:\n(?:          - \S+\n)*)'
    match = re.search(pattern, content)
    if not match:
        print(f"⚠ Could not locate options block in {DIGEST_WF} — skipping workflow update.")
        return

    if f"- {CITY_KEY}" in match.group(0):
        print(f"→ '{CITY_KEY}' already in dispatch options — skipping workflow update.")
        return

    new_entry = f"            - {CITY_KEY}\n"
    updated = re.sub(pattern, lambda m: m.group(0) + new_entry, content)

    with open(DIGEST_WF, "w") as f:
        f.write(updated)

    print(f"→ Added '{CITY_KEY}' to dispatch options in {DIGEST_WF}")


def main() -> None:
    out_path = os.path.join(SOURCES_DIR, f"{CITY_KEY}.yml")
    is_update = os.path.exists(out_path)

    print(f"{'Update' if is_update else 'Add'} City — {CITY_KEY} ({CITY_NAME})")
    print("=" * 50)

    existing = load_existing_sources()
    sources  = discover_sources(existing)
    write_city_file(sources)
    update_digest_workflow()

    print(f"✓ Done. Commit sources/{CITY_KEY}.yml{' and digest.yml' if not is_update else ''} to complete the setup.")


if __name__ == "__main__":
    main()

File: src/add_city.ts
// Creates the two pieces of plumbing a new city needs: a sources/{city}.yml
// skeleton, and an entry in digest.yml's dispatch options.
//
// It deliberately does NOT discover sources. That used to fan out to
// Anthropic, Perplexity and Google and merge the prose each returned, which
// discover.ts later measured as worthless — see its header: Claude wrapped its
// JSON in prose, GPT-5 returned empty output, and neither found anything
// Google's grounded answer had missed. `pnpm discover-sources` does the same
// job in twelve narrow niche calls and reaches the long tail a broad "list
// this city's event sources" question never gets to.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import yaml from "js-yaml";
import { PROJECT_ROOT, requireEnv, SOURCES_ROOT } from "./common.ts";

const CITY_NAME = requireEnv("CITY_NAME");
const CITY_KEY = requireEnv("CITY_KEY");

const DIGEST_WF = join(PROJECT_ROOT, ".github/workflows/digest.yml");

/** The header comment yaml.dump cannot produce, and the centre placeholder a
 * new city must fill in before its first curate. */
const HEADER = `# Event sources for ${CITY_NAME}.
#
# Each entry declares how it is collected:
#   method: scraper — we fetch its listingUrls ourselves (src/adapters/).
#   method: llm     — no verified listing page, so LLM web search covers it.
#
# Every source starts as method: llm. Only probe-sources promotes one, once
# it has verified a listing URL actually yields dated events.
#
# TODO: add this city's centre. Until it is here, curate.ts keeps every event
# and cannot tell a local one from an interstate one (src/locality.ts).
# Deliberately not written as a placeholder: coordinates that are present but
# wrong are worse than absent, since every real venue is then "somewhere else".
# radiusKm must stay smaller than the distance to the nearest other city in
# sources/, or the two will swallow each other's events.
#
# centre:
#   lat: -27.4698
#   lng: 153.0251
#   radiusKm: 50
`;

function writeCityFile(): boolean {
	mkdirSync(SOURCES_ROOT, { recursive: true });
	const outPath = join(SOURCES_ROOT, `${CITY_KEY}.yml`);
	if (existsSync(outPath)) {
		console.log(`→ ${outPath} already exists — left untouched.`);
		return false;
	}

	const cityData = {
		name: CITY_NAME,
		timezone: "Australia/Brisbane",
		sources: { aggregators: [], institutions: [], independents: [] },
	};
	writeFileSync(
		outPath,
		HEADER + yaml.dump(cityData, { noRefs: true, sortKeys: false }),
		"utf-8",
	);
	console.log(`→ Written ${outPath}`);
	return true;
}

function updateDigestWorkflow(): void {
	const content = readFileSync(DIGEST_WF, "utf-8");

	const pattern = /( {8}options:\n(?:( {10}- \S+\n))*)/;
	const match = content.match(pattern);
	if (!match) {
		console.log(
			`⚠ Could not locate options block in ${DIGEST_WF} — skipping workflow update.`,
		);
		return;
	}

	if (match[0].includes(`- ${CITY_KEY}`)) {
		console.log(
			`→ '${CITY_KEY}' already in dispatch options — skipping workflow update.`,
		);
		return;
	}

	const newEntry = `          - ${CITY_KEY}\n`;
	const updated = content.replace(pattern, (m) => m + newEntry);
	writeFileSync(DIGEST_WF, updated, "utf-8");
	console.log(`→ Added '${CITY_KEY}' to dispatch options in ${DIGEST_WF}`);
}

function main(): void {
	console.log(`Add City — ${CITY_KEY} (${CITY_NAME})`);
	console.log("=".repeat(50));

	const created = writeCityFile();
	updateDigestWorkflow();

	if (created) {
		console.log("\nNext:");
		console.log(
			`  1. Set the real centre coordinates in sources/${CITY_KEY}.yml`,
		);
		console.log(
			`  2. pnpm discover-sources --city=${CITY_KEY} --apply   # find sources`,
		);
		console.log(
			`  3. pnpm probe-sources --city=${CITY_KEY} --apply      # promote the scrapable ones`,
		);
	}
	console.log(
		`✓ Done. Commit sources/${CITY_KEY}.yml${created ? " and digest.yml" : ""}.`,
	);
}

main();
